torch/cuda/_memory_viz.py

```diff
@@ -301,6 +309,364 @@ def find_segment(addr):
             format(d)
     return out.getvalue()
 
+def trace_plot(data, device=None, plot_segments=False):
+
+    string_table = []
+    suffix_table = []
+
+    elements = []
+    actions = []
+
+    addr_to_alloc = {}
+
+    initially_allocated = []
+
+    @cache
+    def intern_str(s):
+        string_table.append(s)
+        return len(string_table) - 1
+
+    @cache
+    def intern_suffix(sid, restid):
+        suffix_table.append((sid, restid))
+        return len(suffix_table) - 1
+
+
+    def intern_stack(frames):
+        sids = [intern_str(f) for f in frames]
+        next_id = None
+        for sid in reversed(sids):
+            next_id = intern_suffix(sid, next_id)
+        return next_id
+
+    if device is None:
+        for i, t in enumerate(data['device_traces']):
+            if len(t) > 0:
+                if device is not None:
+                    raise ValueError(f'Both device {device} and {i} have traces, use --device to specify which trace.')
+                device = i
+        if device is None:
+            raise ValueError('No trace information was recorded.')
+
+    trace = data['device_traces'][device]
+
+    def add_element(size, frames, extra=()):
+        frames = [f"{_format_size(size)} allocation", *extra, *(_frame_fmt(f, full_filename=True) for f in frames)]
+        frame_id = intern_stack(frames)
+        elements.append({'size': size, 'info': frame_id})
+        return len(elements) - 1
+
+    if plot_segments:
+        alloc = 'segment_alloc'
+        free = 'segment_free'
+    else:
+        alloc = 'alloc'
+        free = 'free_completed'
+
+    for i, e in enumerate(trace):
+        if e['action'] == alloc:
+            elemid = add_element(e['size'], e['frames'])
+            addr_to_alloc[e['addr']] = elemid
+            actions.append(elemid)
+        elif e['action'] == free:
+            idx = addr_to_alloc.pop(e['addr'], None)
+            if idx is None:
+                idx = add_element(e['size'], e['frames'], extra=('alloc not recorded, stack trace for free:',))
+                initially_allocated.append(idx)
+            actions.append(idx)
+
+    r = {
+        'actions': actions,
+        'elements': elements,
+        'suffix_table': suffix_table,
+        'string_table': string_table,
+        'initially_allocated': list(reversed(initially_allocated)),
+    }
+    plot_data = json.dumps(r)
+    return _memory_over_time_template.replace('$PLOT_DATA', plot_data)
+
+
+_memory_over_time_template = r"""
```

Comment:
For giant templates like this I've found it's better to store them in the native language and then read them in. (e.g. [benchmark cpp template](https://github.com/pytorch/pytorch/blob/master/torch/utils/benchmark/utils/timeit_template.cpp)) That way you get syntax highlighting.

Reply:
I added an inline note here: I agree its better in its own file. Right now I end up download _memory_viz.py to a bunch of places where it would hard to install pytorch (e.g. laptop). I need to figure out a way to make a really small pip installable package for the thing so that it can be multiple files.